src/FilesystemManagement.cpp

```diff
@@ -0,0 +1,365 @@
+            case ENAMETOOLONG:  return Filesystem::ModifyResult::NameTooLong;
+            case EACCES:        return Filesystem::ModifyResult::PermissionDenied;
+            case EPERM:         return Filesystem::ModifyResult::PermissionDenied;
+            case EROFS:         return Filesystem::ModifyResult::ReadOnly;
+            case ENOTDIR:       return Filesystem::ModifyResult::NotADirectory;
+            case EISDIR:        return Filesystem::ModifyResult::IsADirectory;
+            case ENOTEMPTY:     return Filesystem::ModifyResult::NotEmpty;
+            case EIO:           return Filesystem::ModifyResult::IOError;
+            case ENOSPC:        return Filesystem::ModifyResult::NoSpace;
+            case EMLINK:        return Filesystem::ModifyResult::MaxLinksExceeded;
+            case EBUSY:         return Filesystem::ModifyResult::CurrentlyBusy;
+            case ECANCELED:     return Filesystem::ModifyResult::OperationCanceled;
+            default:            return Filesystem::ModifyResult::Unknown;
+        }
+    }
+#endif // MEZZ_Windows
+}
+
+namespace Mezzanine {
+namespace Filesystem {
+    ///////////////////////////////////////////////////////////////////////////////
+    // Basic File Management
+
+    Boole FileExists(const StringView FilePath)
+    {
+    #ifdef MEZZ_Windows
+        std::wstring WidePath = ConvertToWideString(FilePath);
+        DWORD dwAttrib = GetFileAttributesW(WidePath.c_str());
+        return ( dwAttrib != INVALID_FILE_ATTRIBUTES && !( dwAttrib & FILE_ATTRIBUTE_DIRECTORY ) );
+    #else // MEZZ_Windows
+        //return ( ::access(PathAndName.c_str(),F_OK) );
+        struct stat st;
+        if( stat(FilePath.data(),&st) == 0 ) {
+            return S_ISREG(st.st_mode);
+        }
+        return false;
+    #endif // MEZZ_Windows
+    }
+
+    ModifyResult CopyFile(const StringView OldFilePath, const StringView NewFilePath, const Boole FailIfExists)
+    {
+    #ifdef MEZZ_Windows
+        DWORD CopyFlags = COPY_FILE_COPY_SYMLINK;
+        if( FailIfExists ) {
+            CopyFlags |= COPY_FILE_FAIL_IF_EXISTS;
+        }
+        std::wstring WideOldPath = ConvertToWideString(OldFilePath);
+        std::wstring WideNewPath = ConvertToWideString(NewFilePath);
+        return ( ::CopyFileExW(WideOldPath.c_str(),WideNewPath.c_str(),NULL,NULL,NULL,CopyFlags) != 0 ?
+                 ModifyResult::Success :
+                 ConvertErrNo( ::GetLastError() ) );
+    #else // MEZZ_Windows
+        if( FailIfExists && FileExists(NewFilePath.data()) ) {
+            return ModifyResult::AlreadyExists;
+        }
+        std::ifstream SrcStream;
+        std::ofstream DestStream;
+
+        SrcStream.open(String(OldFilePath),std::ios_base::binary | std::ios_base::in);//Converting to String is Ew.
+        DestStream.open(String(NewFilePath),std::ios_base::binary | std::ios_base::out);//Converting to String is Ew.
+
+        DestStream << SrcStream.rdbuf();
+        return ModifyResult::Success;
```

Comment:
Doesn't this work on both platforms?

I suspect link would be faster on the same filesystem, then falling back to this stream implementation otherwise.

https://pubs.opengroup.org/onlinepubs/009695399/functions/link.html

Reply:
This streaming copy could be extracted to its own function `ForceDeepCopy` or something.

Reply:
Hard links are not deep copies.  

And yes, the streaming does work on both platforms, but I'm trying to use OS calls where I can and there isn't an OS call for a straight file copy on posix like there is on Windows.  Hence this solution.  In fact, the cp command does exactly this but in C (according to this source): 
https://stackoverflow.com/a/17666950